Create Dog API Endpoint

parkerra opened the pull request:
## Create Dog Endpoint
Issue Number: #1 

API endpoint has been added to create new dogs in the database. The information about the dog is sent in the POST request body and it verified using Zod. If any of the fields are incorrectly formatted or ObjectIds not present in the database, then a descriptive error is returned from the request.

### Checklist
- [x] Requirements have been implemented
- [x] Acceptance criteria is met
- [x] Database schema docs have been updated or are not necessary
- [x] Code follows design and style guidelines
- [x] Commits follow guidelines (concise, squashed, etc)
- [x] Relevant reviewers (EM) have been assigned to this PR

### How To Test
Send a post request to `http://localhost:3000/api/dogs/` with a JSON body in the format. Note that the ObjectIds and dates are strings and not objects.

{
    "name": "Spots",
    "gender": "Female",
    "breed": "German Shephard",
    "weight": 8,
    "behavior": "No concern",
    "medical": "No concern",
    "other": "No concern",
    "birthOrder": 2,
    "dateOfBirth": "2023-12-09",
    "maternalDemeanor": "Happy",
    "recentLogs": [],
    "parents": ["65039d09f45bcbd64b60ee9a"],
    "instructors": ["6503b1f73944dcb4cfcd8514"],
    "volunteer": "6503b1f73944dcb4cfcd8514",
    "location": "Placed"
}

parkerra committed:
Validate dogs API request fields with Zod

Use Zod to validate and verify fields in the
request body for the dogs API endpoint.

parkerra committed:
Add database operations to create new dog

Operations to create a new dog with input from the
dogs API endpoint have been added

parkerra committed:
Add checks for existing ObjectIDs for createDog

For the createDog database action, checks have been added for
fields with ObjectIds. The createDog method queries the database
for the id in the correct collection for each present field and if
the id is not found in the database then there is an error message.

afazio1 commented on server/db/actions/Dog.js:
```diff
@@ -0,0 +1,57 @@
+export async function createDog(dogData) {
+  await dbConnect();
+
+  // recentLogs array
```
Not sure we need to check if all the logs exist since logs are only created for a dog _after_ it has been created. Just make sure to default `recentLogs` to an empty array in the Zod schema.

afazio1 commented on src/pages/api/dogs/index.js:
```diff
@@ -0,0 +1,90 @@
+    .optional(),
+  dateOfBirth: z.coerce.date(),
+  litterSize: z.number().optional(),
+  birthOrder: z.number().min(1).optional(),
```
I wonder if it is possible to set the max `birthOrder` to `litterSize`

parkerra committed:
Set empty array for default recentLogs

For creating a dog, recentLogs do not need to be specified in
the request body, an empty array will be put into the database
for the dog's recentLogs. Also, the API responses have been
changed from strings to JSON objects.

afazio1 commented on src/pages/api/dogs/index.js:
```diff
@@ -0,0 +1,99 @@
+      .then(() => {
+        return res.status(200).send({
+          success: true,
+          messge: "New dog successfully created!",
```
typo `messge`, let's include `data` as the new dog's `_id` here, and to be consistent we can have a status `201 - Created`

afazio1 commented on src/pages/api/dogs/index.js:
```diff
@@ -0,0 +1,99 @@
+        });
+      });
+  }
+}
```
We should also end the request if they do not use an "allowed" method. Outside the `if (req.method == "POST")` statement, you can add:

```js
res.status(405).send({success: false, message: `Request method ${req.method} is not allowed`})
```

afazio1 commented on server/db/actions/Dog.js:
```diff
@@ -0,0 +1,59 @@
+ * @returns string id of the dog if successfully saved, error otherwise
+ */
+export async function createDog(dogData) {
+  await dbConnect();
```
I would wrap this in a try/catch with a custom error message that is more user friendly. Right now, if this function errors because we are unable to connect to the db, the mongoose error message is thrown (and we don't want the client seeing these dev-specific errors). 

Instead, we can do something like this:
```js
try {
    await dbConnect();
} catch (e) {
    throw new Error("Unable to create dog at this time, please try again");
}
```

afazio1 reviewed (changes requested):
A few minor changes and this PR should be ready to go! Great work 🥳

parkerra committed:
Alter some API responses and check birth order

Some of the API responses have been changed to return data and to
return an unsuccessful state when the user tries to send a request
other than a POST request. Also, now there is a check if the birth order
is greater than the litter size. Cnnecting to the database has
been put inside a try-catch block as well.

afazio1 reviewed (approved):
Looks good
